### PR TITLE
Adding DC/OS support

### PR DESCRIPTION
**Pull request by willgorman**
These are some changes that we've made as part of some work we've been doing to add DC/OS support to Spinnaker.  I'm not sure if this is something that you'd want in marathon-client since it extends beyond just marathon (and we're planning to add more DC/OS APIs, like metronome) However I thought it'd be good to check if this would be acceptable first before we create a fork.

### CODE REVIEW

**Comment by kensipe** [comment]
@willgorman I currently don't have time to review this...   love the support and a quick scan looks good.  hopefully someone from the team picks it up or I'll take a look this weekend.

I want to respond to your metronome comment.  metronome will be going away and will be rolled into marathon.    this client lib will need to support it will it becomes available.   I believe this will happen in the next quarter.

**Comment by ssk2** [comment]
We'll have a review, thanks!

**Comment by colin-msphere** [comment]
We have a team review for this PR scheduled for 2/21.

Thank you for the submission!

File: pom.xml [hunk 1]
```diff
@@ -89,44 +104,44 @@
 				<artifactId>maven-compiler-plugin</artifactId>
 				<version>2.5.1</version>
 				<configuration>
-					<source>1.6</source>
-					<target>1.6</target>
+					<source>1.8</source>
```
Comment (kensipe) [inline pom.xml]: not sure if all envs are ready for  1.8

Comment (kensipe) [inline pom.xml]: jenkins is built with java 1.7 (but we run on 1.8)  need to confirm

File: src/main/java/mesosphere/client/common/HttpResponseException.java [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+        final StringBuilder string = new StringBuilder();
+        if (methodKey != null) {
+            string.append("Error calling ").append(methodKey).append(": ");
+        }
+        string.append(message).append(" (http status: ").append(status)
+                .append(")");
+        if (details != null) {
+            string.append("\n").append("Details: ").append(details);
+        }
+        return string.toString();
```
Comment (kensipe) [inline src/main/java/mesosphere/client/common/HttpResponseException.java]: later `toString` functions are more challenging to read... it would be generally good to use the ToStringBuilder

Comment (kensipe) [inline src/main/java/mesosphere/client/common/HttpResponseException.java]: or the ToStringHelper in guava.

File: src/main/java/mesosphere/client/common/ThrowingSupplier.java [hunk 3]
```diff
@@ -0,0 +1,9 @@
+package mesosphere.client.common;
+
+/**
+ * Version of a {@link java.util.function.Supplier} which is capable of throwing a {@link Throwable}.
+ */
+@FunctionalInterface
```
Comment (kensipe) [inline src/main/java/mesosphere/client/common/ThrowingSupplier.java]: this forces a Java 8 dependency.  need to double check that we can.

File: src/main/java/mesosphere/dcos/client/DCOSClient.java [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+
+import java.io.IOException;
+
+public class DCOSClient {
+
+    public static DCOS getInstance(String endpoint) {
+        return getInstance(endpoint, null);
+    }
+
+    public static DCOS getInstance(String endpoint, final DCOSAuthCredentials authCredentials) {
```
Comment (kensipe) [inline src/main/java/mesosphere/dcos/client/DCOSClient.java]: this is really odd... getInstance(x,y)  calls getInstance(x), which calls get(x,y(null)).   can we clean this up?

File: src/main/java/mesosphere/marathon/client/model/v2/App.java [hunk 5]
```diff
@@ -332,4 +453,47 @@ public String toString() {
 		return ModelUtils.toString(this);
 	}
 
+
+	@Override
+	public boolean equals(final Object o) {
```
Comment (kensipe) [inline src/main/java/mesosphere/marathon/client/model/v2/App.java]: I would love to see this cleaned up... ToStringBuilder or something.   Same with equals and hashcode.

File: src/main/java/mesosphere/marathon/client/model/v2/GetAppNamespaceResponse.java [hunk 6]
```diff
@@ -0,0 +1,28 @@
+package mesosphere.marathon.client.model.v2;
+
+import mesosphere.client.common.ModelUtils;
+
+import com.google.gson.annotations.SerializedName;
+
+import java.util.List;
+
+/**
+ * @author Will Gorman
```
Comment (kensipe) [inline src/main/java/mesosphere/marathon/client/model/v2/GetAppNamespaceResponse.java]: we don't have names on any other classes..  we definitely want to mark you as a contributor in the docs tho!

File: src/main/java/mesosphere/marathon/client/model/v2/GetMetricsResponse.java [hunk 7]
```diff
@@ -0,0 +1,289 @@
...
+        public Integer getCount() {
+            return count;
+        }
+
+        public void setCount(Integer count) {
+            this.count = count;
+        }
+    }
+
+    static class Histogram extends Counter {
```
Comment (kensipe) [inline src/main/java/mesosphere/marathon/client/model/v2/GetMetricsResponse.java]: shouldn't this be available in our metric code somewhere?

File: src/main/java/mesosphere/marathon/client/model/v2/Task.java [hunk 8]
```diff
@@ -2,84 +2,120 @@
...
-	public void setVersion(String version) {
-		this.version = version;
-	}
-
-	@Override
-	public String toString() {
-		return ModelUtils.toString(this);
-	}
+    private String id;
+    private String slaveId;
```
Comment (kensipe) [inline src/main/java/mesosphere/marathon/client/model/v2/Task.java]: I would love to change this to `agentId`.  we are trying to move from slave to agent  naming.

File: src/main/java/mesosphere/marathon/client/model/v2/VersionInfo.java [hunk 9]
```diff
@@ -0,0 +1,29 @@
+package mesosphere.marathon.client.model.v2;
+
+import mesosphere.client.common.ModelUtils;
+
+public class VersionInfo {
```
Comment (kensipe) [inline src/main/java/mesosphere/marathon/client/model/v2/VersionInfo.java]: This is app version info... not to be confused with version info from `/v2/info` which provides marathon version info

File: src/main/java/mesosphere/dcos/client/model/DCOSAuthToken.java [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+                      .plusSeconds(86400)
+                      .isAfter(expiry);
+    }
+
+    public Instant getExpiry() {
+        return expiry;
+    }
+
+    private static Instant extractExpiry(final String token) {
+        // TODO no real error handling here.
```
Comment (kensipe) [inline src/main/java/mesosphere/dcos/client/model/DCOSAuthToken.java]: what should we do here?  I would rather not merge in TODOs

**Comment by colin-msphere** [comment]
I have tested this with our Jenkins plugin and was able to still deploy applications to DC/OS.

File: src/main/java/mesosphere/dcos/client/DCOS.java [hunk 11]
```diff
@@ -0,0 +1,425 @@
...
+import mesosphere.marathon.client.model.v2.GetLeaderResponse;
+import mesosphere.marathon.client.model.v2.GetMetricsResponse;
+import mesosphere.marathon.client.model.v2.GetPluginsResponse;
+import mesosphere.marathon.client.model.v2.GetTasksResponse;
+import mesosphere.marathon.client.model.v2.Group;
+import mesosphere.marathon.client.model.v2.Queue;
+import mesosphere.marathon.client.model.v2.Result;
+
+@Headers({ "Content-Type: application/json", "Accept: application/json" })
+public interface DCOS {
```
Comment (colin-msphere) [inline src/main/java/mesosphere/dcos/client/DCOS.java]: Most of this file are operations against Marathon itself (URL prefix `/marathon/`). Is there a reason these are not in the `Marathon` interface file? Is there  a way to extend `Marathon` to not duplicate `RequestLine`s?

Comment (mtweten) [inline src/main/java/mesosphere/dcos/client/DCOS.java]: I agree, extending the `Marathon` interface would be the way to go - we were just having trouble figuring out a good way to do it.

I believe the problem is that the RequestLines in the `Marathon` interface methods start with the `/v2/` prefix. If we inherit from that interface and then create a `DCOS` instance with an endpoint to a DC/OS master (e.g. http://dcos-master), the inherited `Marathon` methods would resolve to `http://dcos-master/v2/<whatever>`, instead of `http://dcos-master/marathon/v2/<whatever>`.

Basically we couldn't figure out 'built-in' way using feign to prepend that `/marathon/` prefix for inherited marathon methods in the DCOS interface.

Thinking about it a bit now though, it might be possible with a request interceptor for DCOS - if the request's url starts with `/v2`/, we could alter the url and add the `/marathon` prefix

File: pom.xml (hunk 1 above)
Comment (mtweten) [inline pom.xml]: We could definitely target 1.7 if this is a problem.

I believe we only used 1.8 for the convenience methods (e.g. [maybeApp, maybeGroup, maybeSecret](https://github.com/willgorman/marathon-client/blob/1d8712f0aa9ee1115a543500ecc230080eaad411/src/main/java/mesosphere/dcos/client/DCOS.java#L313-L323) which automatically handle 404s) which use the 1.8 Optional and lambdas. Those could be removed, or replaced with Guava's Optional and no lambdas.

File: src/main/java/mesosphere/dcos/client/DCOSClient.java (hunk 4 above)
Comment (willgorman) [inline src/main/java/mesosphere/dcos/client/DCOSClient.java]: Is the desire to avoid needing to create another instance to use with the DCOSAuthTokenHeaderInterceptor or just to reduce the chain of getInstance calls?  Would something like this be preferable?

```java
    public static DCOS getInstance(String endpoint) {
        return buildInstance(endpoint, b -> {});
    }

    public static DCOS getInstance(String endpoint, final DCOSAuthCredentials authCredentials) {
        return buildInstance(endpoint, b ->
                b.requestInterceptor(new DCOSAuthTokenHeaderInterceptor(authCredentials, getInstance(endpoint))));
    }

    private static DCOS buildInstance(String endpoint, Consumer<Feign.Builder> customize) {
        GsonDecoder decoder = new GsonDecoder(ModelUtils.GSON);
        GsonEncoder encoder = new GsonEncoder(ModelUtils.GSON);

        Feign.Builder builder = Feign.builder()
                .encoder(encoder)
                .decoder(decoder)
                .errorDecoder(new DCOSErrorDecoder());
        customize.accept(builder);
        return builder.target(DCOS.class, endpoint);
    }
```

File: src/main/java/mesosphere/dcos/client/DCOS.java [hunk 12]
```diff
@@ -0,0 +1,425 @@
...
+
+    // Miscellaneous
+    @RequestLine("GET /marathon/ping")
+    String getPing() throws DCOSException;
+
+    @RequestLine("GET /marathon/metrics")
+    GetMetricsResponse getMetrics() throws DCOSException;
+
+    // Convenience methods for identifiable resources.
+    default Optional<App> maybeApp(final String id) throws DCOSException {
```
Comment (TrevinTeacutter) [inline src/main/java/mesosphere/dcos/client/DCOS.java]: Just noting here we cannot use all of these default methods with Java 1.7

File: src/main/java/mesosphere/dcos/client/DCOS.java [hunk 13]
```diff
@@ -0,0 +1,425 @@
...
+
+    default Optional<Secret> maybeSecret(final String secretStore, final String secretPath) throws DCOSException {
+        return resource(() -> getSecret(secretStore, secretPath));
+    }
+
+    default Optional<GetAppNamespaceResponse> maybeApps(final String namespace) throws DCOSException {
+        return resource(() -> getApps(namespace));
+    }
+
+    default DeleteAppTasksResponse deleteAppTasksWithTaskId(String appId, String taskId, boolean force) throws DCOSException {
```
Comment (TrevinTeacutter) [inline src/main/java/mesosphere/dcos/client/DCOS.java]: Just noting here we cannot use all of these default methods with Java 1.7

File: src/main/java/mesosphere/dcos/client/DCOS.java (hunk 11 above)
Comment (colin-msphere) [inline src/main/java/mesosphere/dcos/client/DCOS.java]: Is there a way to create a `Marathon` client from the `DCOS` client and alter the server address to include the `/marathon` suffix? It may be easier to alter the target (server) than the `RequestLine` path.

File: pom.xml (hunk 1 above)
Comment (colin-msphere) [inline pom.xml]: I would prefer to keep this to 1.7, due to its use within [a Jenkins plugin](https://github.com/jenkinsci/marathon-plugin).

File: src/main/java/mesosphere/dcos/client/DCOS.java (hunk 11 above)
Comment (TrevinTeacutter) [inline src/main/java/mesosphere/dcos/client/DCOS.java]: So talking with Will and Tweten offline, there is potentially another option.

We could add custom headers to all of the `Marathon` interface routes and just write a `RequestInterceptor` in the DCOS client that looks for those headers and adjusts the path appropriately. This also allows for further extension if other APIs are made available through DCOS (like `Metronome`, if it wasn't getting swallowed into `Marathon`).

**Comment by willgorman** [comment]
Is the 1.7 support a definite requirement?  If not, we've addressed all the other comments if you want to take another look.  If it is, then we can make that change as well.

**Comment by colin-msphere** [comment]
After some discussion, the 1.7 support is not a requirement. We can move forward with 1.8.

File: src/main/java/mesosphere/marathon/client/model/v2/App.java [hunk 14]
```diff
@@ -151,11 +192,11 @@ public void setContainer(Container container) {
 		this.container = container;
 	}
 
-	public Map<String, String> getEnv() {
+	public Map<String, Object> getEnv() {
```
Comment (kensipe) [inline src/main/java/mesosphere/marathon/client/model/v2/App.java]: @willgorman I don't understand the switch from `<String, String>` to `<String, Object>`.   Why is this necessary or useful?

Comment (willgorman) [inline src/main/java/mesosphere/marathon/client/model/v2/App.java]: This change was made to work with secrets since an environment variable in the json can be either a string or an object containing the path of the secret to inject for that variable.  We could introduce a type that contains both a string and a secret and that serializes as only one or the other. However we went this route since it is at least somewhat passive in that existing code that creates an App should be unaffected. If there's a better option that's entirely passive we'd prefer to do that.

Comment (willgorman) [inline src/main/java/mesosphere/marathon/client/model/v2/App.java]: @kensipe How would you prefer that we proceed with this? The options I see are:
* Leave it as is
* Introduce a string/secret union-ish type
* Avoid making any non-passive change to the App class.  I'm not sure there's a good way to do that without creating a separate App in the DCOS package and copying large chunks of the API as well.
